File: src/app/src/main/java/com/huawei/touchmenot/java/hms/ui/MLFaceGraphic.java
/*
 * Copyright 2020. Huawei Technologies Co., Ltd. All rights reserved.
 *
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */
package com.huawei.touchmenot.java.hms.ui;

import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.Typeface;

import com.huawei.hms.mlsdk.common.MLPosition;
import com.huawei.hms.mlsdk.face.MLFace;
import com.huawei.hms.mlsdk.face.MLFaceKeyPoint;
import com.huawei.hms.mlsdk.face.MLFaceShape;
import com.huawei.touchmenot.java.hms.camera.GraphicOverlay;
import com.huawei.touchmenot.R;
import com.huawei.touchmenot.java.main.common.Constants;

import java.text.DecimalFormat;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;

public class MLFaceGraphic extends GraphicOverlay.Graphic {
    private static final float BOX_STROKE_WIDTH = 8.0f;
    private static final float LINE_WIDTH = 5.0f;

    private final GraphicOverlay overlay;

    private final Paint facePositionPaint;
    private final Paint landmarkPaint;
    private final Paint boxPaint;

    private final Paint facePaint;
    private final Paint eyePaint;
    private final Paint eyebrowPaint;
    private final Paint lipPaint;
    private final Paint nosePaint;
    private final Paint noseBasePaint;
    private final Paint textPaint;
    private final Paint probilityPaint;
    private static String getColour_1 = "#ffcc66", getColour_2 = "#00ccff", getColour_3 = "#006666",
            getColor_4 = "#ffff00", getColor_5 = "#ff6699", getColor_6 = "#990000";
    private int TEXT_SIZE_24 = 24, TEXT_SIZE_35 = 35;

    private float start = 350f, width = 500f, VALUE_POINT_5F = 0.5f;

    private volatile MLFace mFace;

    public MLFaceGraphic(GraphicOverlay overlay, MLFace face) {
        super(overlay);
        mFace = face;
        this.overlay = overlay;
        final int selectedColor = Color.WHITE;

        facePositionPaint = new Paint();
        facePositionPaint.setColor(selectedColor);

        textPaint = new Paint();
        textPaint.setColor(Color.WHITE);
        textPaint.setTextSize(TEXT_SIZE_24);
        textPaint.setTypeface(Typeface.DEFAULT);

        probilityPaint = new Paint();
        probilityPaint.setColor(Color.WHITE);
        probilityPaint.setTextSize(TEXT_SIZE_35);
        probilityPaint.setTypeface(Typeface.DEFAULT);

        landmarkPaint = new Paint();
        landmarkPaint.setColor(Color.RED);
        landmarkPaint.setStyle(Paint.Style.FILL);
        landmarkPaint.setStrokeWidth(Constants.RADIUS_10F);

        boxPaint = new Paint();
        boxPaint.setColor(Color.WHITE);
        boxPaint.setStyle(Paint.Style.STROKE);
        boxPaint.setStrokeWidth(BOX_STROKE_WIDTH);

        facePaint = new Paint();
        facePaint.setColor(Color.parseColor(getColour_1));
        facePaint.setStyle(Paint.Style.STROKE);
        facePaint.setStrokeWidth(LINE_WIDTH);

        eyePaint = new Paint();
        eyePaint.setColor(Color.parseColor(getColour_2));
        eyePaint.setStyle(Paint.Style.STROKE);
        eyePaint.setStrokeWidth(LINE_WIDTH);

        eyebrowPaint = new Paint();
        eyebrowPaint.setColor(Color.parseColor(getColour_3));
        eyebrowPaint.setStyle(Paint.Style.STROKE);
        eyebrowPaint.setStrokeWidth(LINE_WIDTH);

        nosePaint = new Paint();
        nosePaint.setColor(Color.parseColor(getColor_4));
        nosePaint.setStyle(Paint.Style.STROKE);
        nosePaint.setStrokeWidth(LINE_WIDTH);

        noseBasePaint = new Paint();
        noseBasePaint.setColor(Color.parseColor(getColor_5));
        noseBasePaint.setStyle(Paint.Style.STROKE);
        noseBasePaint.setStrokeWidth(LINE_WIDTH);

        lipPaint = new Paint();
        lipPaint.setColor(Color.parseColor(getColor_6));
        lipPaint.setStyle(Paint.Style.STROKE);
        lipPaint.setStrokeWidth(LINE_WIDTH);
    }

    public List<String> sortHashMap(HashMap<String, Float> map) {
        Set<Map.Entry<String, Float>> entey = map.entrySet();
        List<Map.Entry<String, Float>> list = new ArrayList<Map.Entry<String, Float>>(entey);
        Collections.sort(list, new Comparator<Map.Entry<String, Float>>() {
            @Override
            public int compare(Map.Entry<String, Float> o1, Map.Entry<String, Float> o2) {
                if (o2.getValue() - o1.getValue() >= Constants.INIT_ZERO) {
                    return 1;
                } else {
                    return -1;
                }
            }
        });
        List<String> emotions = new ArrayList<>();
        for (int i = Constants.INIT_ZERO; i < Constants.INIT_TWO; i++) {
            emotions.add(list.get(i).getKey());
        }
        return emotions;
    }

    @Override
    public void draw(Canvas canvas) {
        if (mFace == null) {
            return;
        }

        float x = start;
        float y = overlay.getHeight() - Constants.RADIUS_300F;
        HashMap<String, Float> emotions = new HashMap<>();
        emotions.put(Constants.STR_SMILING, mFace.getEmotions().getSmilingProbability());
        emotions.put(Constants.STR_NEUTRAL, mFace.getEmotions().getNeutralProbability());
        emotions.put(Constants.STR_ANGRY, mFace.getEmotions().getAngryProbability());
        emotions.put(Constants.STR_FEAR, mFace.getEmotions().getFearProbability());
        emotions.put(Constants.STR_SAD, mFace.getEmotions().getSadProbability());
        emotions.put(Constants.STR_DISGUST, mFace.getEmotions().getDisgustProbability());
        emotions.put(Constants.STR_SURPRISE, mFace.getEmotions().getSurpriseProbability());
        List<String> result = sortHashMap(emotions);

        DecimalFormat decimalFormat = new DecimalFormat(String.valueOf(R.string.DECIMAL_PATTERN));
        // Draw the facial feature value.
        canvas.drawText(R.string.Left_Eye + decimalFormat.format(mFace.getFeatures().getLeftEyeOpenProbability()), x, y, probilityPaint);
        x = x + width;
        canvas.drawText(R.string.Right_Eye + decimalFormat.format(mFace.getFeatures().getRightEyeOpenProbability()), x, y, probilityPaint);
        y = y - Constants.RADIUS_40F;
        x = start;
        canvas.drawText(R.string.Moutstache_Probability + decimalFormat.format(mFace.getFeatures().getMoustacheProbability()), x, y, probilityPaint);
        x = x + width;
        canvas.drawText(R.string.GLASS_PROBABILITY + decimalFormat.format(mFace.getFeatures().getSunGlassProbability()), x, y, probilityPaint);
        y = y - Constants.RADIUS_40F;
        x = start;
        canvas.drawText(R.string.HAT + decimalFormat.format(mFace.getFeatures().getHatProbability()), x, y, probilityPaint);
        x = x + width;
        canvas.drawText(String.valueOf(R.string.AGE) + mFace.getFeatures().getAge(), x, y, probilityPaint);
        y = y - Constants.RADIUS_40F;
        x = start;
        String sex = (mFace.getFeatures().getSexProbability() > VALUE_POINT_5F) ? Constants.STR_GENDER_FEMALE : Constants.STR_GENDER_MALE;
        canvas.drawText(R.string.GENDER + sex, x, y, probilityPaint);
        x = x + width;
        canvas.drawText(R.string.EULER_ANGLE_Y + decimalFormat.format(mFace.getRotationAngleY()), x, y, probilityPaint);
        y = y - Constants.RADIUS_40F;
        x = start;
        canvas.drawText(R.string.EULER_ANGEL_Z + decimalFormat.format(mFace.getRotationAngleZ()), x, y, probilityPaint);
        x = x + width;
        canvas.drawText(R.string.EULER_ANGLE_X + decimalFormat.format(mFace.getRotationAngleX()), x, y, probilityPaint);
        y = y - Constants.RADIUS_40F;
        x = start;
        canvas.drawText(result.get(Constants.INIT_ZERO), x, y, probilityPaint);

        // Draw a face contour.
        if (mFace.getFaceShapeList() != null) {
            for (MLFaceShape faceShape : mFace.getFaceShapeList()) {
                if (faceShape == null) {
                    continue;
                }
                List<MLPosition> points = faceShape.getPoints();
                for (int i = Constants.INIT_ZERO; i < points.size(); i++) {
                    MLPosition point = points.get(i);
                    canvas.drawPoint(translateX(point.getX().floatValue()), translateY(point.getY().floatValue()), boxPaint);
                    if (i != (points.size() - Constants.INIT_ONE)) {
                        MLPosition next = points.get(i + Constants.INIT_ONE);
                        if (point != null && point.getX() != null && point.getY() != null) {
                            if (i % Constants.INIT_THREE == Constants.INIT_ZERO) {
                                canvas.drawText(i + Constants.INIT_ONE + Constants.STR_EMPTY, translateX(point.getX().floatValue()), translateY(point.getY().floatValue()), textPaint);
                            }
                            canvas.drawLines(new float[]{translateX(point.getX().floatValue()), translateY(point.getY().floatValue()),
                                    translateX(next.getX().floatValue()), translateY(next.getY().floatValue())}, getPaint(faceShape));
                        }
                    }
                }
            }
        }
        // Face Key Points
        for (MLFaceKeyPoint keyPoint : mFace.getFaceKeyPoints()) {
            if (keyPoint != null) {
                MLPosition point = keyPoint.getPoint();
                canvas.drawCircle(
                        translateX(point.getX()),
                        translateY(point.getY()),
                        Constants.RADIUS_10F, landmarkPaint);
            }
        }
    }

    private Paint getPaint(MLFaceShape faceShape) {
        switch (faceShape.getFaceShapeType()) {
            case MLFaceShape.TYPE_LEFT_EYE:
            case MLFaceShape.TYPE_RIGHT_EYE:
                return eyePaint;
            case MLFaceShape.TYPE_BOTTOM_OF_LEFT_EYEBROW:

            case MLFaceShape.TYPE_BOTTOM_OF_RIGHT_EYEBROW:
            case MLFaceShape.TYPE_TOP_OF_LEFT_EYEBROW:
            case MLFaceShape.TYPE_TOP_OF_RIGHT_EYEBROW:
                return eyebrowPaint;
            case MLFaceShape.TYPE_BOTTOM_OF_LOWER_LIP:
            case MLFaceShape.TYPE_TOP_OF_LOWER_LIP:
            case MLFaceShape.TYPE_BOTTOM_OF_UPPER_LIP:
            case MLFaceShape.TYPE_TOP_OF_UPPER_LIP:
                return lipPaint;
            case MLFaceShape.TYPE_BOTTOM_OF_NOSE:
                return noseBasePaint;
            case MLFaceShape.TYPE_BRIDGE_OF_NOSE:
                return nosePaint;
            default:
                return facePaint;
        }
    }
}

File: src/app/src/main/java/com/huawei/touchmenot/kotlin/hms/ui/MLFaceGraphic.kt
/*
 * Copyright 2020. Huawei Technologies Co., Ltd. All rights reserved.
 *
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */
package com.huawei.touchmenot.kotlin.hms.ui

import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.Typeface
import com.huawei.hms.mlsdk.face.MLFace
import com.huawei.hms.mlsdk.face.MLFaceShape
import com.huawei.touchmenot.kotlin.hms.camera.GraphicOverlay
import com.huawei.touchmenot.kotlin.hms.camera.GraphicOverlay.Graphic
import com.huawei.touchmenot.R
import com.huawei.touchmenot.kotlin.main.common.Constants
import java.text.DecimalFormat
import java.util.*

class MLFaceGraphic(private val overlay: GraphicOverlay, private val mFace: MLFace?) : Graphic(overlay) {
    private val facePositionPaint: Paint
    private val landmarkPaint: Paint
    private val boxPaint: Paint
    private val facePaint: Paint
    private val eyePaint: Paint
    private val eyebrowPaint: Paint
    private val lipPaint: Paint
    private val nosePaint: Paint
    private val noseBasePaint: Paint
    private val textPaint: Paint
    private val probilityPaint: Paint
    private val TEXT_SIZE_24 = 24
    private val TEXT_SIZE_35 = 35
    private val start = 350f
    private val width = 500f
    private val VALUE_POINT_5F = 0.5f

    fun sortHashMap(map: HashMap<String, Float>): List<String> {
        val entey: Set<Map.Entry<String, Float>> = map.entries
        val list: List<Map.Entry<String, Float>> = ArrayList(entey)
        Collections.sort(list) { o1, o2 ->
            if (o2.value - o1.value >= Constants.INIT_ZERO) {
                1
            } else {
                -1
            }
        }
        val emotions: MutableList<String> = ArrayList()
        for (i in Constants.INIT_ZERO until Constants.INIT_TWO) {
            emotions.add(list[i].key)
        }
        return emotions
    }

    override fun draw(canvas: Canvas) {
        if (mFace == null) {
            return
        }
        var x = start
        var y = overlay.height - Constants.RADIUS_300F
        val emotions = HashMap<String, Float>()
        emotions[Constants.STR_SMILING] = mFace.emotions.smilingProbability
        emotions[Constants.STR_NEUTRAL] = mFace.emotions.neutralProbability
        emotions[Constants.STR_ANGRY] = mFace.emotions.angryProbability
        emotions[Constants.STR_FEAR] = mFace.emotions.fearProbability
        emotions[Constants.STR_SAD] = mFace.emotions.sadProbability
        emotions[Constants.STR_DISGUST] = mFace.emotions.disgustProbability
        emotions[Constants.STR_SURPRISE] = mFace.emotions.surpriseProbability
        val result = sortHashMap(emotions)
        val decimalFormat = DecimalFormat(R.string.DECIMAL_PATTERN.toString())
        // Draw the facial feature value.
        canvas.drawText(R.string.Left_Eye.toString() + decimalFormat.format(mFace.features.leftEyeOpenProbability.toDouble()), x, y, probilityPaint)
        x = x + width
        canvas.drawText(R.string.Right_Eye.toString() + decimalFormat.format(mFace.features.rightEyeOpenProbability.toDouble()), x, y, probilityPaint)
        y = y - Constants.RADIUS_40F
        x = start
        canvas.drawText(R.string.Moutstache_Probability.toString() + decimalFormat.format(mFace.features.moustacheProbability.toDouble()), x, y, probilityPaint)
        x = x + width
        canvas.drawText(R.string.GLASS_PROBABILITY.toString() + decimalFormat.format(mFace.features.sunGlassProbability.toDouble()), x, y, probilityPaint)
        y = y - Constants.RADIUS_40F
        x = start
        canvas.drawText(R.string.HAT.toString() + decimalFormat.format(mFace.features.hatProbability.toDouble()), x, y, probilityPaint)
        x = x + width
        canvas.drawText(R.string.AGE.toString() + mFace.features.age, x, y, probilityPaint)
        y = y - Constants.RADIUS_40F
        x = start
        val sex = if (mFace.features.sexProbability > VALUE_POINT_5F) Constants.STR_GENDER_FEMALE else Constants.STR_GENDER_MALE
        canvas.drawText(R.string.GENDER.toString() + sex, x, y, probilityPaint)
        x = x + width
        canvas.drawText(R.string.EULER_ANGLE_Y.toString() + decimalFormat.format(mFace.rotationAngleY.toDouble()), x, y, probilityPaint)
        y = y - Constants.RADIUS_40F
        x = start
        canvas.drawText(R.string.EULER_ANGEL_Z.toString() + decimalFormat.format(mFace.rotationAngleZ.toDouble()), x, y, probilityPaint)
        x = x + width
        canvas.drawText(R.string.EULER_ANGLE_X.toString() + decimalFormat.format(mFace.rotationAngleX.toDouble()), x, y, probilityPaint)
        y = y - Constants.RADIUS_40F
        x = start
        canvas.drawText(result[Constants.INIT_ZERO], x, y, probilityPaint)

        // Draw a face contour.
        if (mFace.faceShapeList != null) {
            for (faceShape in mFace.faceShapeList) {
                if (faceShape == null) {
                    continue
                }
                val points = faceShape.points
                for (i in Constants.INIT_ZERO until points.size) {
                    val point = points[i]
                    canvas.drawPoint(translateX(point!!.x.toFloat()), translateY(point.y.toFloat()), boxPaint)
                    if (i != points.size - Constants.INIT_ONE) {
                        val next = points[i + Constants.INIT_ONE]
                        if (point != null && point.x != null && point.y != null) {
                            if (i % Constants.INIT_THREE == Constants.INIT_ZERO) {
                                canvas.drawText((i + Constants.INIT_ONE).toString(), translateX(point.x.toFloat()), translateY(point.y.toFloat()), textPaint)
                            }
                            canvas.drawLines(floatArrayOf(translateX(point.x.toFloat()), translateY(point.y.toFloat()),
                                    translateX(next.x.toFloat()), translateY(next.y.toFloat())), getPaint(faceShape))
                        }
                    }
                }
            }
        }
        // Face Key Points
        for (keyPoint in mFace.faceKeyPoints) {
            if (keyPoint != null) {
                val point = keyPoint.point
                canvas.drawCircle(
                        translateX(point.x),
                        translateY(point.y),
                        Constants.RADIUS_10F, landmarkPaint)
            }
        }
    }

    private fun getPaint(faceShape: MLFaceShape): Paint {
        return when (faceShape.faceShapeType) {
            MLFaceShape.TYPE_LEFT_EYE, MLFaceShape.TYPE_RIGHT_EYE -> eyePaint
            MLFaceShape.TYPE_BOTTOM_OF_LEFT_EYEBROW, MLFaceShape.TYPE_BOTTOM_OF_RIGHT_EYEBROW, MLFaceShape.TYPE_TOP_OF_LEFT_EYEBROW, MLFaceShape.TYPE_TOP_OF_RIGHT_EYEBROW -> eyebrowPaint
            MLFaceShape.TYPE_BOTTOM_OF_LOWER_LIP, MLFaceShape.TYPE_TOP_OF_LOWER_LIP, MLFaceShape.TYPE_BOTTOM_OF_UPPER_LIP, MLFaceShape.TYPE_TOP_OF_UPPER_LIP -> lipPaint
            MLFaceShape.TYPE_BOTTOM_OF_NOSE -> noseBasePaint
            MLFaceShape.TYPE_BRIDGE_OF_NOSE -> nosePaint
            else -> facePaint
        }
    }

    companion object {
        private const val BOX_STROKE_WIDTH = 8.0f
        private const val LINE_WIDTH = 5.0f
        private const val getColour_1 = "#ffcc66"
        private const val getColour_2 = "#00ccff"
        private const val getColour_3 = "#006666"
        private const val getColor_4 = "#ffff00"
        private const val getColor_5 = "#ff6699"
        private const val getColor_6 = "#990000"
    }

    init {
        val selectedColor = Color.WHITE
        facePositionPaint = Paint()
        facePositionPaint.color = selectedColor
        textPaint = Paint()
        textPaint.color = Color.WHITE
        textPaint.textSize = TEXT_SIZE_24.toFloat()
        textPaint.typeface = Typeface.DEFAULT
        probilityPaint = Paint()
        probilityPaint.color = Color.WHITE
        probilityPaint.textSize = TEXT_SIZE_35.toFloat()
        probilityPaint.typeface = Typeface.DEFAULT
        landmarkPaint = Paint()
        landmarkPaint.color = Color.RED
        landmarkPaint.style = Paint.Style.FILL
        landmarkPaint.strokeWidth = Constants.RADIUS_10F
        boxPaint = Paint()
        boxPaint.color = Color.WHITE
        boxPaint.style = Paint.Style.STROKE
        boxPaint.strokeWidth = BOX_STROKE_WIDTH
        facePaint = Paint()
        facePaint.color = Color.parseColor(getColour_1)
        facePaint.style = Paint.Style.STROKE
        facePaint.strokeWidth = LINE_WIDTH
        eyePaint = Paint()
        eyePaint.color = Color.parseColor(getColour_2)
        eyePaint.style = Paint.Style.STROKE
        eyePaint.strokeWidth = LINE_WIDTH
        eyebrowPaint = Paint()
        eyebrowPaint.color = Color.parseColor(getColour_3)
        eyebrowPaint.style = Paint.Style.STROKE
        eyebrowPaint.strokeWidth = LINE_WIDTH
        nosePaint = Paint()
        nosePaint.color = Color.parseColor(getColor_4)
        nosePaint.style = Paint.Style.STROKE
        nosePaint.strokeWidth = LINE_WIDTH
        noseBasePaint = Paint()
        noseBasePaint.color = Color.parseColor(getColor_5)
        noseBasePaint.style = Paint.Style.STROKE
        noseBasePaint.strokeWidth = LINE_WIDTH
        lipPaint = Paint()
        lipPaint.color = Color.parseColor(getColor_6)
        lipPaint.style = Paint.Style.STROKE
        lipPaint.strokeWidth = LINE_WIDTH
    }
}